experimental/listener.go

```diff
@@ -0,0 +1,64 @@
+package experimental
+
+import (
+	"context"
+
+	"github.com/tetratelabs/wazero/api"
+)
+
+// FunctionListenerFactoryKey is a context.Context Value key. Its associated value should be a FunctionListenerFactory.
+//
+// Note: This is interpreter-only for now!
+// See https://github.com/tetratelabs/wazero/issues/451
+type FunctionListenerFactoryKey struct{}
+
+// FunctionListenerFactory returns FunctionListeners to be notified when a function is called.
+type FunctionListenerFactory interface {
+	// NewListener returns a FunctionListener for a defined function. If nil is returned, no
+	// listener will be notified.
+	NewListener(FunctionDefinition) FunctionListener
+}
+
+// FunctionListener can be registered for any function via FunctionListenerFactory to
+// be notified when the function is called.
+type FunctionListener interface {
+	// Before is invoked before a function is called. ctx is the context of the caller function.
+	// The returned context will be used as the context of this function call. To add context
+	// information for this function call, add it to ctx and return the updated context. If
+	// no context information is needed, return ctx as is.
+	Before(ctx context.Context, paramValues []uint64) context.Context
+
+	// After is invoked after a function is called. ctx is the context of this function call.
+	// The err parameter is nil on success.
+	After(ctx context.Context, err error, resultValues []uint64)
+}
+
+// TODO: We need to add tests to enginetest to ensure contexts nest. A good test can use a combination of call and call
+// indirect in terms of depth and breadth. The test could show a tree 3 calls deep where the there are a couple calls at
+// each depth under the root. The main thing this can help prevent is accidentally swapping the context internally.
+
+// TODO: Errors aren't handled, and the After hook should accept one along with the result values.
```

Comment:
`After` looks like it has an `error` now so guess this was already DOed?

Reply:
It has a parameter, but "errors aren't handled" is still true. There's neither code nor tests to do that.